x/auth/keeper/migrations.go

```diff
@@ -0,0 +1,146 @@
+package keeper
+
+import (
+	"fmt"
+
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+
+	"github.com/cosmos/cosmos-sdk/x/auth/vesting/exported"
+
+	"github.com/gogo/protobuf/grpc"
+	"github.com/gogo/protobuf/proto"
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+)
+
+const (
+	delegatorDelegationPath = "/cosmos.staking.v1beta1.Query/DelegatorDelegations"
+	balancesPath            = "/cosmos.bank.v1beta1.Query/AllBalances"
+)
+
+// Migrator is a struct for handling in-place store migrations.
+type Migrator struct {
+	keeper      AccountKeeper
+	queryServer grpc.Server
+}
+
+// NewMigrator returns a new Migrator.
+func NewMigrator(keeper AccountKeeper, queryServer grpc.Server) Migrator {
+	return Migrator{keeper: keeper, queryServer: queryServer}
+}
+
+// Migrate1to2 migrates from version 1 to 2.
+func (m Migrator) Migrate1to2(ctx sdk.Context) error {
+	m.keeper.IterateAccounts(ctx, func(account types.AccountI) (stop bool) {
+		asVesting := vesting(account)
+		if asVesting == nil {
+			return false
+		}
+
+		addr := account.GetAddress().String()
+		balance := getBalance(
+			ctx,
+			addr,
+			m.queryServer,
+		)
+
+		delegations := getDelegatorDelegations(
+			ctx,
+			addr,
+			m.queryServer,
+		)
+
+		if delegations.IsAllGTE(asVesting.GetOriginalVesting()) {
+			delegations = asVesting.GetOriginalVesting()
+		}
+
+		if balance.IsAllLT(delegations) {
+			balance = balance.Add(delegations...)
+		}
```

Comment:
Why are you doing these fixups before using `TrackDelegation`?

The first one is invalid for accounts that have delegated non-vesting tokens - tokens that have been earned via rewards or sent to the account and then delegated.  See the failing test that I added here:

https://github.com/ovrclk/cosmos-sdk/blob/8d2cc6cc36d7ffc66d3f1d5efc1d65345ad4655e/x/auth/keeper/migrations_test.go#L242-L267

I'm not sure what the second one is supposed to do (`if balance is less than delegations, add delegations to balance`?)

I still think it'd be easier to just set the values to what they should be.
```
DelegatedVesting = min(delegations,vesting)
DelegatedFree    = delegations - DelegatedVesting
```

Reply:
Here is an account that has not had any new delegations since v0.39

```
$ akash query account akash1ksfwd5h3tcd9kk8seftpavvdfdgkvvv2y7gvya
'@type': /cosmos.vesting.v1beta1.PeriodicVestingAccount
base_vesting_account:
  base_account:
    account_number: "35"
    address: akash1ksfwd5h3tcd9kk8seftpavvdfdgkvvv2y7gvya
    pub_key:
      '@type': /cosmos.crypto.secp256k1.PubKey
      key: Akf9otCknYTAn9s5qDO3BkdIL5wos+p86jRm/YbLj6se
    sequence: "6149"
  delegated_free:
  - amount: "18127450000"
    denom: uakt
  delegated_vesting:
  - amount: "66666670000"
    denom: uakt
  end_time: "1664114400"
  original_vesting:
  - amount: "66666670000"
    denom: uakt
start_time: "1601042400"
vesting_periods:
- amount:
  - amount: "33333335000"
    denom: uakt
  length: "31536000"
- amount:
  - amount: "16666667500"
    denom: uakt
  length: "15638400"
- amount:
  - amount: "16666667500"
    denom: uakt
  length: "15897600"
$ akash query staking delegations akash1ksfwd5h3tcd9kk8seftpavvdfdgkvvv2y7gvya
delegation_responses:
- balance:
    amount: "84794120000"
    denom: uakt
  delegation:
    delegator_address: akash1ksfwd5h3tcd9kk8seftpavvdfdgkvvv2y7gvya
    shares: "84794120000.000000000000000000"
    validator_address: akashvaloper1ksfwd5h3tcd9kk8seftpavvdfdgkvvv2wux25h
pagination:
  next_key: null
  total: "0"
$ akash query bank balances akash1ksfwd5h3tcd9kk8seftpavvdfdgkvvv2y7gvya
balances:
- amount: "35284933"
  denom: uakt
pagination:
  next_key: null
  total: "0"
```

The export of the previous chain had the same values.